File: 2015/Day05_2015.py
niceStrings = 0
niceStrings2 = 0

with open("Day05_2015.txt") as file:
    for line in file:
        vowels = 0
        dblLetter = False
        dblOtherLetter = False
        badStrings = False
        repeatedPair = False
        prevChar =''
        prevChar2 = ''

        for char in line:
            if (char == 'a') or (char == 'e') or (char == 'i') or (char == 'o') or (char == 'u'):
                vowels += 1
            if char == prevChar:
                dblLetter = True
            if char == prevChar2:
                dblOtherLetter = True
            if ((prevChar + char) == 'ab') or ((prevChar + char) == 'cd') or ((prevChar + char) == 'pq') or ((prevChar + char) == 'xy'):
                badStrings = True
            prevChar2 = prevChar
            prevChar = char

        if (vowels >= 3) and (dblLetter == True) and (badStrings == False):
            niceStrings += 1
        
        for i in range(len(line)-3):
            for j in range(i+2, len(line)-1):
                if line[i:i+2] == line[j:j+2]:
                    repeatedPair = True

        if (dblOtherLetter == True and repeatedPair == True):
            niceStrings2 += 1

file.close()

print("The number of nice strings in Santa's text file: " + str(niceStrings))
print("The number of nice strings under the new rules: " + str(niceStrings2))

File: 2015/Day05_2015.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <cstring>

int main()
{
    std::ifstream file("Day05_2015.txt");
    std::string line;

    int niceStrings = 0;
    int niceStrings2 = 0;

    while (std::getline(file, line))
    {
        int vowels = 0;
        bool dblLetter = false;
        bool dblOtherLetter = false;
        bool badStrings = false;
        bool repeatedPair = false;
        char prevChar = '\0';
        char prevChar2 = '\0';
        
        for (int i = 0; i < line.size(); i++)
        {
            if (line[i] == 'a' || line[i] == 'e' || line[i] == 'i' || line[i] == 'o' || line[i] == 'u')
                vowels++;
            if (line[i] == prevChar)
                dblLetter = true;
            if (line[i] == prevChar2)
                dblOtherLetter = true;

            std::string lastTwoChars{prevChar, line[i]};

            if (lastTwoChars == "ab" || lastTwoChars == "cd" || lastTwoChars == "pq" || lastTwoChars == "xy")
                badStrings = true;

            prevChar2 = prevChar;
            prevChar = line[i];
        }

        if (vowels >= 3 && dblLetter == true && badStrings == false)
            niceStrings++;

        for (int i = 0; i < line.size() - 3; i++)
        {
            for (int j = i + 2; j < line.size() - 1; j++)
            {
                if (line[i] == line[j] && line[i+1] == line[j+1])
                    repeatedPair = true;
            }
        }

        if (dblOtherLetter == true && repeatedPair == true)
            niceStrings2++;
    }

    file.close();

    std::cout << "The number of nice strings in Santa's text file: " << niceStrings << "\n";
    std::cout << "The number of nice strings under the new rules: " << niceStrings2;

    return 0;
}